src/bytecode/scanner.rs

```diff
@@ -1,312 +0,0 @@
-use anyhow::{anyhow, bail, Result};
-
-#[derive(Debug)]
-pub struct Scanner {
-    source: Vec<char>, // TODO: this is not ideal, we should be pointing back to the original source string to save memory
-    start: usize,      // The start of the current lexeme
-    current: usize,    // The current character under consideration
-    line: usize,
-}
-
-impl Scanner {
-    fn new(source: &str) -> Scanner {
-        Scanner {
-            source: source.chars().collect(),
-            start: 0,
-            current: 0,
-            line: 1,
-        }
-    }
-
-    fn make_token(&self, typ: TokenType) -> Token {
-        Token {
-            typ,
-            start: self.start,
-            length: self.current - self.start,
-            line: self.line,
-        }
-    }
-
-    fn is_at_end(&self) -> bool {
```

Comment:
This differs from the conventional formulation, which puts the offset at the char not **at** but **after** the last char.
```java
  private boolean isAtEnd() {
    return current >= source.length();
  }
  ```

Is this intentional? I can't get a positive read.

Reply:
I think I had the convention different than in the book for some reason here - maybe because I was using a `Vec<char>` in this implementation? But this is all deleted - the new implementation in this RP doesn't need this at all, I just rely on the underlying iterator to tell me when it is exhausted.